### ISSUE TITLE
Adversarial Robustness for an MNIST network

### ISSUE THREAD

**Issue by cherrywoods** [NeuralNetworkVerification/Marabou#494]
Hi,
I was following [this example](https://neuralnetworkverification.github.io/Marabou/Examples/5_DisjunctionConstraintExample.html) with the aim in mind of proving or disproving the robustness of a MNIST neural network for a specific input image. 

In the attached script, I am analyzing a copy of the 5x100 ReLU network from the [ETH Robustness Analyzer (ERAN)](https://github.com/eth-sri/eran), looking for adversarial examples for a specific image of the digit nine. 

Unfortunately when running the script, Marabou outputs that the network is robust for that specific input, while I have evidence that it is not (adversarial example discovered using ERAN contained in the script).

I really can't get behind what I am doing wrong. Maybe you could comment on that?
My code (the script is an excerpt) works fine for ACAS Xu properties. 

Script and ONNX network are contained in this ZIP file: [code-and-network.zip](https://github.com/NeuralNetworkVerification/Marabou/files/7747171/code-and-network.zip)
The script downloads the MNIST training set in order to obtain the image using torchvision. I hope that the order of the dataset is maintained across downloads. I did not include the image because I am unsure about the legal side of uploading individual MNIST samples.

You will need the following pip packages to run the script:
```
torchvision==0.10.0+cpu numpy~=1.19
```
and an installation of Marabou. 

Best regards,
David

**Comment by persistz** [comment]
+1 
I also found dozens of similar cases on MNIST dataset. 
Looking forward to a fix for the problem.

**Comment by anwu1219** [comment]
Hi @cherrywoods 

Thanks for bringing this up. I was able to reproduce the error.
A couple of issues here:
1. In the python API, currently it's hard to differ between a runtime error and "unsat". The unsat that you saw was actually an runtime error. I have submitted a PR (https://github.com/NeuralNetworkVerification/Marabou/pull/511) which refactors the python API (long overdue) to fix this issue.
2. The error comes from the disjunction constraint, which probably contains a bug that needs to be fixed. I created a separate issue for this.
3. For now, I would suggest do what many other NN verifiers do: iterate through each adversarial label and check individually whether a counter-example exists. Please find attached the code for checking your example property. 
 [code-and-network.zip](https://github.com/NeuralNetworkVerification/Marabou/files/7989698/code-and-network.zip)
The solver was able to find a counter-example with the script running on the branch of that PR (hopefully it will be merge soon). Please let me know if the code does not make sense.

**Comment by cherrywoods** [comment]
Thank you for looking into this and thanks for providing the fix! The code makes perfect sense. 

Can you suggest a strategy what to do when the result is "unknown"? I suspect such issues to come up, because I have encountered such cases in experiments with ERAN, which uses a similar mechanism, as you mentioned.

**Comment by anwu1219** [comment]
Hi @cherrywoods ,

Great! 
In the code that I sent you, the only scenario where we print out "unknown: is when the solver timed out to some user-specified limit. I would expect only "TIMEOUT", "sat", and "unsat". If something else comes up as return code, please create an issue. 

That said, you can impose the timeout by setting the timeoutInSeconds option in the createOptions() method:
https://github.com/NeuralNetworkVerification/Marabou/blob/2f15d95a148c73919e20e3ee67a0e022ca9fa13f/maraboupy/Marabou.py#L121-L127
The createOptions method returns an option object to pass into the MarabouNetwork.solve() method. Please check this out for an example:
https://github.com/NeuralNetworkVerification/Marabou/blob/2f15d95a148c73919e20e3ee67a0e022ca9fa13f/maraboupy/test/test_network.py#L100-L114

Like in ERAN, this timeout does not include the abstraction-based preprocessing time (e.g., running DeepPoly analysis). Unlike ERAN, by default there is no timeout.

For performance evaluation, I would build with Gurobi on (https://github.com/NeuralNetworkVerification/Marabou#use-lp-relaxation) and try the following 2 configurations:
`createOptions(imeoutInSeconds=..., solveWithMILP=False, skipLpTighteningAfterSplit=True)`
`createOptions(timeoutInSeconds=..., solveWithMILP=True, skipLpTighteningAfterSplit=True)`

Please let me know if you encountered issues!

**Comment by cherrywoods** [comment]
Hi @anwu1219 
Thanks for your elaborate reply. I was misreading the code. 

My question actually refers to the following: When we get a "SAT" result, we can not conclude for sure that the whole disjunction is also SAT. 
When we find a satisfying assignment in the code that assignment is for y9 <= yi. Yet it may be that some other clause of the disjunction actually holds for that assignment. 
That issue comes up for me with ERAN sometimes and there is no handling mechanism in ERAN. I was wondering whether you could suggest some a mechanism for handling it in Marabou.

The reason why I am concerned with this is that I need proper counterexamples for my application.

**Comment by anwu1219** [comment]
I'm not sure I fully understand. 
The disjunction is satisfied as long as at least one of its disjunct is satisfied. 
Therefore, as long as the solver returns "sat" for any one of the 9 checks, the original disjunction must be satisfied.
For example, if y9 <= y3 for some assignment, perhaps it is true that y9 <= y4 and y4 >= y3 for that assignment, but it doesn't mean the disjunction is not satisfied. 

Perhaps you're interested in finding a counter-example that will be classified as a specific target label? This can be done by adding the following constraints, which literally says "the target label is the maximal".

```
for other_label in range(10):
    if other_label != target_label:
        eq = MarabouUtils.Equation(MarabouCore.Equation.LE)
        eq.addAddend(1, output_variables[other_label])
        eq.addAddend(-1, output_variables[target_label])
        eq.setScalar(0)
        networkTemp.addEquation(eq)
```
If the solver returns 'sat', then it has found a counter-example that satisfies the given constraints.
Again you can iterate through each target label and do this check.

**Comment by cherrywoods** [comment]
Hi @anwu1219,
Thank you once more for your answer. You are of course entirely correct. As you may have noticed I am currently not immersed in the subject. I thought I had checked my previous question well enough, but no, clearly not. 

The issue that I wanted to describe actually doesn't occur for the specification that I used in the example. 
When using a specification that contains conjunctions and disjunctions, sometimes in ERAN you obtain improper (spurious) counterexamples. 

I think the issue can't come up here now. For sure it can't come up with the specification I have provided. 
I can't tell at the moment whether it can come up here for more complex specifications, I think it can't. 

Thank you for your patience! Thank you as well for the code and the suggestions for configurations.

Best regards,
David
